packages/protocol/contracts/identity/FederatedAttestations.sol

```diff
@@ -113,12 +109,69 @@ contract FederatedAttestations is
-   *   for an identifier produced by a list of issuers
+   * @notice Helper function for _lookupAttestations to calculate the
+             total number of attestations completed for an identifier
+             by each trusted issuer, from unrevoked signers only
+   * @param identifier Hash of the identifier
+   * @param trustedIssuers Array of n issuers whose attestations will be included
+   * @return [0] Sum total of attestations found
+   *         [1] Array of number of attestations found per issuer
+   */
+  function getNumUnrevokedAttestations(bytes32 identifier, address[] memory trustedIssuers)
+    internal
+    view
+    returns (uint256, uint256[] memory)
+  {
+    uint256 totalAttestations = 0;
+    uint256[] memory countsPerIssuer = new uint256[](trustedIssuers.length);
+
+    OwnershipAttestation[] memory attestationsPerIssuer;
+    for (uint256 i = 0; i < trustedIssuers.length; i = i.add(1)) {
+      attestationsPerIssuer = identifierToAttestations[identifier][trustedIssuers[i]];
+      for (uint256 j = 0; j < attestationsPerIssuer.length; j = j.add(1)) {
+        if (!revokedSigners[attestationsPerIssuer[j].signer]) {
+          totalAttestations = totalAttestations.add(1);
+          countsPerIssuer[i] = countsPerIssuer[i].add(1);
+        }
+      }
+    }
+    return (totalAttestations, countsPerIssuer);
+  }
+
+  /**
+   * @notice Helper function for _lookupAttestations to calculate the
+             total number of attestations completed for an identifier
+             by each trusted issuer
+   * @param identifier Hash of the identifier
+   * @param trustedIssuers Array of n issuers whose attestations will be included
+   * @return [0] Sum total of attestations found
+   *         [1] Array of number of attestations found per issuer
+   */
+  function getNumAttestations(bytes32 identifier, address[] memory trustedIssuers)
+    internal
+    view
+    returns (uint256, uint256[] memory)
+  {
+    uint256 totalAttestations = 0;
+    uint256 numAttestationsForIssuer;
+    uint256[] memory countsPerIssuer = new uint256[](trustedIssuers.length);
+
+    for (uint256 i = 0; i < trustedIssuers.length; i = i.add(1)) {
+      numAttestationsForIssuer = identifierToAttestations[identifier][trustedIssuers[i]].length;
+      totalAttestations = totalAttestations.add(numAttestationsForIssuer);
+      countsPerIssuer[i] = numAttestationsForIssuer;
+    }
+    return (totalAttestations, countsPerIssuer);
+  }
+
+  /**
+   * @notice Returns info about up to `maxAttestations` attestations for
+   *   `identifier` produced by unrevoked signers of `trustedIssuers`
    * @param identifier Hash of the identifier
    * @param trustedIssuers Array of n issuers whose attestations will be included
    * @param maxAttestations Limit the number of attestations that will be returned
-   * @return for m found attestations, m <= maxAttestations:
+   * @return [0] Array of number of attestations returned per issuer
```

Comment:
IIUC, this array of counts is intended to provide information to the caller to resolve the issuer for each attestations. To me, this seems like a somewhat unintuitive way to do that. The obvious way, would be to include an array of issuers the same way signers are indicated. A way to compress that would be to return a `uint8[] memory` array of indices into the `trustedIssuers` struct. As long as the result is <= 32 attestations (seems likely to me) then this would only take up one word. (It would cap the maximum number of trusted issuers to 256, but that seems like a reasonable cap).

Of course, these two approaches could be combined by returning a `uint8[] memory` array of counts for each issuer, but that seems to me like a needless optimization at that point, unless you expect each identifier to have many attestations (i.e. > 32). This is especially the case because any one account can only have one attestation per issuer.

If you want to take return data compression further, I might suggest removing timestamp and signer return values, which are unlikely to be used by most callers. You can instead add an index return as a `uint8[] memory` array that would allow the caller to lookup the signer or timestamp by directly indexing into `identifierToAttestations`. (We would need to add a function to allow users to do this direct indexing, but that is easy). (Note that my suggestions of using `uint8` for the index caps the number of attestations per `(issuer, identifier)` to 256. I would be surprised if this is too small, but we could alternatively us a larger index type.)

Reply:
> IIUC, this array of counts is intended to provide information to the caller to resolve the issuer for each attestations

Yes -- this follows how this is done in `Attestations.sol`, in `batchGetAttestationStats`. Wondering if you think consistency with that is worth maintaining? Otherwise, I think returning a uint8[] memory array of trustedIssuer indices would  be clean as well! As a side note, in either case, we can make the SDK wrappers for this more intuitive (i.e. not return just the matches, but probably return a dictionary with attestations keyed by issuer, for ex).

> You can instead add an index return as a uint8[] memory array that would allow the caller to lookup the signer or timestamp by directly indexing into identifierToAttestations. (We would need to add a function to allow users to do this direct indexing, but that is easy)

Agreed on the signer -- I think the timestamp (whether issuedOn or publishedOn) could reasonably be desired by callers to determine recency of attestations. I guess the main downside here is requiring many separate calls to get batch attestation information; do you think this is worthwhile tradeoff? I don't have a strong opinion here though, and don't have great intuition on whether this tradeoff is good one for users calling via the SDK vs. contracts calling this, wdyt?

For adding a function -- since it's a public storage mapping I think this accessor already exists by default (if the user knows the index), unless I'm misunderstanding what you mean?